### ISSUE TITLE
Podman unable to resolve DNS queries(?) — aka pulling images does not work

### ISSUE THREAD

**Issue by jwflory** [containers/libpod#1948]
**Is this a BUG REPORT or FEATURE REQUEST?**:

[//]: # **kind bug**

**Description**

Pulling new images does not work. Podman returns `no such host` for all of the registries it tries to hit. This issue does not happen with Docker on the same system.

I also tried running with SELinux set to permissive, since I had similar issues [once before](https://bugzilla.redhat.com/show_bug.cgi?id=1585802) with Docker. However, there was no change with SELinux set to permissive.

**Steps to reproduce the issue:**

1. podman run --rm -it fedora bash

**Describe the results you received:**

```
Trying to pull docker.io/fedora:latest...Failed
Trying to pull registry.fedoraproject.org/fedora:latest...Failed
Trying to pull quay.io/fedora:latest...Failed
Trying to pull registry.access.redhat.com/fedora:latest...Failed
Trying to pull registry.centos.org/fedora:latest...Failed
unable to pull fedora: 5 errors occurred:

* Error determining manifest MIME type for docker://fedora:latest: pinging docker registry returned: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: no such host
* Error determining manifest MIME type for docker://registry.fedoraproject.org/fedora:latest: pinging docker registry returned: Get https://registry.fedoraproject.org/v2/: dial tcp: lookup registry.fedoraproject.org: no such host
* Error determining manifest MIME type for docker://quay.io/fedora:latest: pinging docker registry returned: Get https://quay.io/v2/: dial tcp: lookup quay.io: no such host
* Error determining manifest MIME type for docker://registry.access.redhat.com/fedora:latest: pinging docker registry returned: Get https://registry.access.redhat.com/v2/: dial tcp: lookup registry.access.redhat.com: no such host
* Error determining manifest MIME type for docker://registry.centos.org/fedora:latest: pinging docker registry returned: Get https://registry.centos.org/v2/: dial tcp: lookup registry.centos.org: no such host
```

**Describe the results you expected:**

```
[root@hostname /] # 
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

I'm using the official Fedora package.

```
podman version 0.10.1.3
```

**Output of `podman info`:**

```
host:
  BuildahVersion: 1.5-dev
  Conmon:
    package: podman-0.10.1.3-1.gitdb08685.fc28.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.12.0-dev, commit: 4a03e555ee4105308fa4d814d4ce02a059af0b7f-dirty'
  Distribution:
    distribution: fedora
    version: "28"
  MemFree: 10145501184
  MemTotal: 16693481472
  OCIRuntime:
    package: runc-1.0.0-59.dev.gitccb5efd.fc28.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.1-dev'
  SwapFree: 8388603904
  SwapTotal: 8388603904
  arch: amd64
  cpus: 4
  hostname: fossbook.justinwflory.com
  kernel: 4.19.5-200.fc28.x86_64
  os: linux
  uptime: 3h 57m 47.64s (Approximately 0.12 days)
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ContainerStore:
    number: 0
  GraphDriverName: vfs
  GraphOptions: []
  GraphRoot: /home/jflory/.local/share/containers/storage
  GraphStatus: {}
  ImageStore:
    number: 0
  RunRoot: /run/user/1000/run
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Bare-metal Fedora 28 Workstation (4.19.5-200.fc28.x86_64).

**Comment by mheon** [comment]
What are the contents of `/etc/resolve.conf`?
Can you pull images with other Go programs (`buildah`, `skopeo`)?

**Comment by baude** [comment]
Do you have an `/etc/containers/registries.conf` ?  And if so, what is inside it. Also, does podman --log-level=debug reveal anything interesting? also, mind updating to the latest podman version? you are reporting against an old version.

**Comment by mheon** [comment]
@baude That is the latest on F28 for now, until we can get 0.12.1 out the door tomorrow. Assuming CI doesn't decide to take a vacation again

**Comment by jwflory** [comment]
> **@mheon**:
> What are the contents of /etc/resolve.conf?

It looks like this:

```
# Generated by NetworkManager
search justinwflory.com
nameserver 192.168.1.1
```

For debugging, I added `nameserver 1.1.1.1`, but nothing changed.

> **@baude**:
> Do you have an /etc/containers/registries.conf ? And if so, what is inside it.

Yes, see below.

```
[registries.search]
registries = ['docker.io', 'registry.fedoraproject.org', 'quay.io', 'registry.access.redhat.com', 'registry.centos.org']

[registries.insecure]
registries = []

[registries.block]
registries = []
```

> **@baude**:
> Also, does podman --log-level=debug reveal anything interesting?

I'm not sure if this is helpful or not. I stopped copying after the first host failed for readability's sake, since the rest were duplicates of this error:

```
INFO[0000] running as rootless                          
DEBU[0000] Not configuring container store              
INFO[0000] Found CNI network podman (type=bridge) at /etc/cni/net.d/87-podman-bridge.conflist 
DEBU[0000] Initializing boltdb state at /home/jflory/.local/share/containers/storage/libpod/bolt_state.db 
DEBU[0000] Set libpod namespace to ""                   
WARN[0000] AppArmor security is not available in rootless mode 
DEBU[0000] Using bridge netmode                         
INFO[0000] running as rootless                          
DEBU[0000] [graphdriver] trying provided driver "vfs"   
INFO[0000] Found CNI network podman (type=bridge) at /etc/cni/net.d/87-podman-bridge.conflist 
DEBU[0000] Initializing boltdb state at /home/jflory/.local/share/containers/storage/libpod/bolt_state.db 
DEBU[0000] Set libpod namespace to ""                   
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]docker.io/library/ubuntu:latest" 
DEBU[0000] reference "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]docker.io/library/ubuntu:latest" does not resolve to an image ID 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]docker.io/library/ubuntu:latest" 
DEBU[0000] reference "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]docker.io/library/ubuntu:latest" does not resolve to an image ID 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]localhost/ubuntu:latest" 
DEBU[0000] reference "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]localhost/ubuntu:latest" does not resolve to an image ID 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]docker.io/library/ubuntu:latest" 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]registry.fedoraproject.org/ubuntu:latest" 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]quay.io/ubuntu:latest" 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]registry.access.redhat.com/ubuntu:latest" 
DEBU[0000] parsed reference into "[vfs@/home/jflory/.local/share/containers/storage+/run/user/1000/run]registry.centos.org/ubuntu:latest" 
Trying to pull docker.io/ubuntu:latest...DEBU[0000] Using registries.d directory /etc/containers/registries.d for sigstore configuration 
DEBU[0000]  Using "default-docker" configuration        
DEBU[0000]  No signature storage configuration found for docker.io/library/ubuntu:latest 
DEBU[0000] Looking for TLS certificates and private keys in /etc/docker/certs.d/docker.io 
DEBU[0000] GET https://registry-1.docker.io/v2/         
DEBU[0000] Ping https://registry-1.docker.io/v2/ err Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: no such host (&url.Error{Op:"Get", URL:"https://registry-1.docker.io/v2/", Err:(*net.OpError)(0xc4200b4320)}) 
DEBU[0000] GET https://registry-1.docker.io/v1/_ping    
DEBU[0000] Ping https://registry-1.docker.io/v1/_ping err Get https://registry-1.docker.io/v1/_ping: dial tcp: lookup registry-1.docker.io: no such host (&url.Error{Op:"Get", URL:"https://registry-1.docker.io/v1/_ping", Err:(*net.OpError)(0xc4200b44b0)}) 
DEBU[0000] Error pulling image ref //ubuntu:latest: Error determining manifest MIME type for docker://ubuntu:latest: pinging docker registry returned: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: no such host 
Failed
```

> **@baude**:
> also, mind updating to the latest podman version? you are reporting against an old version.

Once the 0.12.1.1 package update in [Bodhi](https://bodhi.fedoraproject.org/updates/FEDORA-2018-e8e19475eb) pushes to the testing repos for F28, I can re-test.

**Comment by mheon** [comment]
Alright, you clearly have a valid nameserver, and yet just as clearly the Go code can't make DNS queries.

@mtrmac It seems like we're failing in c/image here, but I'm pretty sure none of the image code does things that would affect DNS resolution

@jwflory Anything unusual about your setup? I'm pretty sure this is a Go standard library HTTP request failing, which is very well tested and should be basically foolproof for a simple query like this, so I'm thinking we might be seeing some sort of environment issue.

**Comment by jwflory** [comment]
> **_@mheon_**:
> @jwflory Anything unusual about your setup?

I can't guess what. I don't work with Golang often so I haven't made many changes. The only thing I've consciously changed about my Golang environment is setting a custom `$GOPATH` somewhere else in my user's home directory.

I also had a similar issue in Fedora 27 with Docker ([bug 1585802](https://bugzilla.redhat.com/show_bug.cgi?id=1585802)). However, the problem there was SELinux. Reinstalling some packages let Docker resolve DNS again. I did set SELinux to permissive with Podman when debugging for this ticket, but nothing changed then either.

**Comment by baude** [comment]
Does pulling a fully qualified image work?
i.e.

```
podman pull registry.fedoraproject.org/fedora:29
```

**Comment by baude** [comment]
@jwflory is this still a problem?

**Comment by jwflory** [comment]
@baude I haven't had a chance to run this again on the affected machine. I'll try to update soon and try again ASAP. Sorry for the slow response, I am taking semester exams this week.

**Comment by jwflory** [comment]
Thanks for your patience. I was able to re-test now.

> @baude:
> also, mind updating to the latest podman version? you are reporting against an old version.

I updated to 0.12.1.1 and this bug is still present.

> @baude:
> Does pulling a fully qualified image work?

Unfortunately, still no dice:

```
$ podman pull registry.fedoraproject.org/fedora:29

Trying to pull registry.fedoraproject.org/fedora:29...Failed
error pulling image "registry.fedoraproject.org/fedora:29": unable to pull registry.fedoraproject.org/fedora:29: unable to pull image: Error determining manifest MIME type for docker://registry.fedoraproject.org/fedora:29: pinging docker registry returned: Get https://registry.fedoraproject.org/v2/: dial tcp: lookup registry.fedoraproject.org: no such host
```

**Comment by baude** [comment]
this has me baffled.  is that same user able to resolve names to IP addresses normally?

**Comment by jwflory** [comment]
@baude In a non-container context? I'm able to successfully resolve DNS with `host` in the command line.

**Comment by rhatdan** [comment]
You are running podman inside of a container?  Or directly on the host?  
Are you doing this with non-root podman?
Does sudo podman pull 
Work for you ?

**Comment by jwflory** [comment]
> You are running podman inside of a container? Or directly on the host?

I am running it directly on my host.

>  Are you doing this with non-root podman?

Yes, I am running as my user.

> Does `sudo podman pull` work for you ?

Yes, actually this worked perfectly:

```
$ sudo podman pull fedora
Trying to pull docker.io/fedora:latest...Getting image source signatures
Copying blob sha256:cd6c8343b59020c6bef022d3f81902a428b0552f1b3c0938fea36383eb68ff6c
 85.64 MB / 85.64 MB [=====================================================] 10s
Copying config sha256:25e6809f6fabf364c645430961bb3a02adc5b93f7bcdc8a41b692100cd0aa8bd
 2.20 KB / 2.20 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
25e6809f6fabf364c645430961bb3a02adc5b93f7bcdc8a41b692100cd0aa8bd
```

**Comment by rhatdan** [comment]
And podman pull fails when you run as non root.

**Comment by jwflory** [comment]
> And podman pull fails when you run as non root.

Yes, it fails.

**Comment by rhatdan** [comment]
jwflory could you attempt to wget off of that host?

If you pull an image from docker.io does it work?

**Comment by mtrmac** [comment]
To isolate this a bit more, please create a `foo.go` with the following contents:
```go
package main

import (
	"fmt"
	"net/http"
	"net/http/httputil"
)

func main() {
	resp, err := http.Get("https://registry.fedoraproject.org/v2/")
	fmt.Printf("Error: %v\n", err)
	if err == nil {
		dr, _ := httputil.DumpResponse(resp, true)
		fmt.Printf("Body:\n===\n%s\n===\n", string(dr))
	}
}
```
and check whether `go run foo.go` fails/succeeds in the same way as `podman pull registry.fedoraproject.org/fedora:latest`; if it does, that’s a much smaller program to check, and `strace` might be useful for pinpointing where the behavior differs.

**Comment by jwflory** [comment]
> jwflory could you attempt to wget off of that host?

`wget` works as expected on my host with my user.

> If you pull an image from docker.io does it work?

Yes, I am able to pull an image from docker.io with `docker` CLI as my user without root (I am in the `docker` group). A similar issue was once reported with Docker in [bug 1585802](https://bugzilla.redhat.com/show_bug.cgi?id=1585802).

> To isolate this a bit more, please create a foo.go and check whether go run foo.go fails/succeeds in the same way as podman pull registry.fedoraproject.org/fedora:latest

Seems like it worked?

```
$ go run foo.go 
Error: <nil>
Body:
===
HTTP/2.0 200 OK
Content-Length: 2
Accept-Ranges: bytes
Age: 29
Appserver: proxy09.fedoraproject.org
Apptime: D=1062
Content-Type: application/json; charset=utf-8
Date: Thu, 03 Jan 2019 21:03:06 GMT
Docker-Distribution-Api-Version: registry/2.0
Referrer-Policy: same-origin
Server: Apache/2.4.34 (Fedora)
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
Vary: Accept
Via: 1.1 varnish (Varnish/5.1)
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Varnish: 16105005 16985940
X-Xss-Protection: 1; mode=block

{}
===
```

**Comment by mtrmac** [comment]
> Seems like it worked? I can't tell

Yes.  That’s even more surprising. I guess it’s one of:

- We are doing something to affect the DNS lookup inside the code? I can’t see what that would be.
- There is a difference between the `go` environments compiling the test program and `podman` (different versions, CGO disabled/enabled, …)
- There is something specific about the `podman` binary on the target system (different SELinux context)

IIRC you said that `podman pull` breaks but `sudo podman pull` works, so tracking down where that behavior diverges (perhaps via strike) could be a next step forward.

**Comment by jwflory** [comment]
> There is something specific about the `podman` binary on the target system (different SELinux context)

The current context on `/usr/bin/podman` is: `system_u:object_r:container_runtime_exec_t:s0`

> IIRC you said that `podman pull` breaks but `sudo podman pull` works, so tracking down where that behavior diverges (perhaps via strike) could be a next step forward.

Let me know if I can do anything to help test further.

**Comment by rhatdan** [comment]
Maybe an strace would help, could you attach that.  
Are you able to get this to happen on other machines?  If you create a new clean account does it work.

sudu sh
useradd test1
su - test1
podman pull registry.fedoraproject.org/fedora:29

**Comment by jwflory** [comment]
> Maybe an strace would help, could you attach that.

This is my first time using `strace` for debugging like this. What's the best way for me to use it that is most helpful for you all?

> Are you able to get this to happen on other machines? If you create a new clean account does it work.

On another machine originally installed with Fedora 28, I do not have this issue. This laptop was originally a F23/F24 install. It's on my list to eventually do a fresh F29 install sometime in February when I have time.

Looks like the same problem with the clean account though:

```
[root@fossbook ~]# useradd test-podman
[root@fossbook ~]# su - test-podman 
[test-podman@fossbook ~]$ podman pull registry.fedoraproject.org/fedora:29
Trying to pull registry.fedoraproject.org/fedora:29...Failed
error pulling image "registry.fedoraproject.org/fedora:29": unable to pull registry.fedoraproject.org/fedora:29: unable to pull image: Error determining manifest MIME type for docker://registry.fedoraproject.org/fedora:29: pinging docker registry returned: Get https://registry.fedoraproject.org/v2/: dial tcp: lookup registry.fedoraproject.org: no such host
```

**Comment by rhatdan** [comment]
If you do `buildah from registry.fedoraproject.org/fedora:29`, does that succeed?

It seems like name resolution is getting screwed up for a single command?

**Comment by rhatdan** [comment]
This has been fixed in buildah. We need similar fixes in podman.

@baude Could you take a look at the fixes in buildah.

**Comment by jwflory** [comment]
Oh awesome, glad to hear this was fixed in `buildah`. Just curious… what did the problem end up being? It's been hard for me to reproduce in any environment except the one where I face this bug.

**Comment by rhatdan** [comment]
@jwflory We had to do a major rewrite/simplification of buildha pull.  There were many issues in parsing that had leaked all through the code.  We cleaned them all up.

**Comment by rhatdan** [comment]
Sadly no one has worked on this.  @mtrmac Do you have any time to work on this to clean it up.

**Comment by mtrmac** [comment]
@rhatdan If you know this has been fixed in `Buildah`, please point out the fixes.

I have no idea what is going on, and apart from randomly diffing `strace`, I have no other way to move forward without a reproducer I can try locally.

**Comment by rhatdan** [comment]
@mtrmac I have no idea, All I know is we did a major rework on buildah pull and would like to see similar done to podman pull.  For the specific Issue, I have no idea if this is fixed in buildah yet.

**Comment by jwflory** [comment]
For what it's worth, this issue was resolved on my "problem" machine. I can pull images and I haven't done a fresh install yet either.

**Comment by rhatdan** [comment]
Could have been some network setup problems on this machine?

**Comment by StefanD986** [comment]
Here is maybe some help for anyone who runs into this issue. In my situation I was running podman on Debian Bookworm in a (vagrant managed) Virtualbox VM. I had exactly the issue as the reporter of the issue (`podman pull` not working with error `dial tcp: lookup registry-1.docker.io on 10.0.2.3:53: no such host`, but DNS lookups, e.g. via `host` or `nslookup` worked fine). 

I solved the issue by enabling the *Host's Resolver as a DNS Proxy in NAT Mode* using `vboxmanage modifyvm "debian" --natdnshostresolver1 on`, or in the `Vagrantfile` with 

```ruby
config.vm.provider "virtualbox" do |v|
  ... # other configuration

  v.customize [
    "modifyvm",
    :id,
    "--natdnshostresolver1",
    "on"
  ]
end
```

References:
I found the solution here: https://serverfault.com/questions/453185/vagrant-virtualbox-dns-10-0-2-3-not-working
some more details about the DNS resovler configuration options https://www.virtualbox.org/manual/ch09.html#nat-adv-dns
